### PR TITLE
Improved Line Detection 

### PR DESCRIPTION
**Pull request by hardik01shah**
1. Add band sampling to the MLP in `gluefactory/models/lines/pold2_extractor.py` and `gluefactory/models/lines/pold2_mlp.py`. Earlier, MLP used samples along the candidate line from DF and AF. We now ammend that to use samples across a band (parallelogram) between candidate keypoint pairs.
2. Add hierarchical weak filtering (discard candidate keypoint pairs using the weak filter by gradually increasing the number of sampled values along the line) in `gluefactory/models/lines/pold2_extractor.py`
3. Update computation of DF and AF to use JPLDD instead of DeepLSD for generation of dataset for MLP training in `gluefactory/datasets/pold2_mlp_dataset.py`.
4. Refactor all configs in `gluefactory/configs`
5. Add a testbed python file that can be used to visualize and debug JPLDD outputs, intermediate outputs of line detection and output lines in `notebooks/jpldd_linedet_testbed.py`. Edit [commit: 006ff38c7d6eec7d1dc1154b6296c24fff7ca0ef]: Add side-by-side visualization with DeepLSD outputs for debugging and comparison
6. Update `gluefactory/train.py` to use the `use_gpu` argument to configure the device for training.
7. Add NMS post line detection [Refer to the PR: https://github.com/FelixMoeller3/glue-factory/pull/15]

### CODE REVIEW

**Comment by FelixMoeller3** [comment]
Thx for the Pr, will review tomorrow evening :)

File: gluefactory/models/lines/pold2_extractor.py [hunk 1]
```diff
@@ -128,10 +130,35 @@ def get_coordinates(self, points, indices, coeffs, coeffs_second):
...
+            return indices_image
+        lines = points[indices_image]
+        diff = lines[:, 0] - lines[:, 1]
+        diff = torch.sum(diff ** 2, dim=1)
+
+        return indices_image[diff > (min_line_length**2)]
+
+    def filter_large_lines(self, points, indices_image):
+        """
+        Filter out lines that are too short.
```
Comment (FelixMoeller3) [inline gluefactory/models/lines/pold2_extractor.py:148]: We need to change this comment haha

**Review by FelixMoeller3 — APPROVED** [review]

File: gluefactory/models/extractors/joint_point_line_extractor.py [hunk 2]
```diff
@@ -197,14 +194,23 @@ def _init(self, conf):
...
             logger.warning(f"Load model parameters from checkpoint {conf.checkpoint}")
             chkpt = torch.load(conf.checkpoint, map_location=torch.device("cpu"))
+
+            # remove mlp weights from line detection
+            chkpt["model"] = {k: v for k, v in chkpt["model"].items() if not ("mlp" in k)}
+
             self.load_state_dict(
-                chkpt["model"], strict=False
+                chkpt["model"], strict=True
             )  # set to false as otherwise error raised for missing mlp weights
```
Comment (RaphaelKreft) [inline gluefactory/models/extractors/joint_point_line_extractor.py:203]: if you change to strict=True, please adapt the comment, to why strict is used

File: gluefactory/train.py [hunk 3]
```diff
@@ -73,6 +73,7 @@
     "pr_curves": {},
     "plot": None,
     "submodules": [],
+    "use_gpu": True
```
Comment (RaphaelKreft) [inline gluefactory/train.py:76]: what happens if this is true and cuda is not available?

File: notebooks/jpldd_linedet_testbed.py [hunk 4]
```diff
@@ -0,0 +1,250 @@
...
+    ax[2].set_title('Angle Field')
+    ax[2].imshow(get_flow_vis(distance_field, angle_field))
+
+    ax[3].axis('off')
+    ax[3].set_title('Original')
+    ax[3].imshow(img.permute(1,2,0))
+    ax[3].scatter(keypoints[:,1],keypoints[:,0], marker="o", color="red", s=3)
+
+# Set Output Directory
+DEBUG_DIR = "tmp_testbed"
```
Comment (RaphaelKreft) [inline notebooks/jpldd_linedet_testbed.py:64]: Const vars should be at the top

File: notebooks/jpldd_linedet_testbed.py [hunk 5]
```diff
@@ -0,0 +1,250 @@
...
+if torch.cuda.is_available():
+    device = 'cuda'
+elif torch.backends.mps.is_built():
+    device = 'mps'
+else:
+    device = 'cpu'
+print(f"Device Used: {device}")
+
+## DeepLSD Model
+deeplsd_conf = {
```
Comment (RaphaelKreft) [inline notebooks/jpldd_linedet_testbed.py:80]: Also please move model configs to the top. This is more practical for users of this script

**Review by RaphaelKreft — CHANGES_REQUESTED** [review]
good overall, some nitpicks to address

File: gluefactory/train.py (hunk 3 above)
Comment (FelixMoeller3) [reply gluefactory/train.py:76]: Then [this](https://github.com/FelixMoeller3/glue-factory/blob/abccb0e33984a699630310d140e6313352331eac/gluefactory/train.py#L237) assert fails

File: gluefactory/models/lines/pold2_extractor.py (hunk 1 above)
Comment (hardik01shah) [reply gluefactory/models/lines/pold2_extractor.py]: done

File: gluefactory/models/extractors/joint_point_line_extractor.py (hunk 2 above)
Comment (hardik01shah) [reply gluefactory/models/extractors/joint_point_line_extractor.py]: done

File: notebooks/jpldd_linedet_testbed.py (hunk 4 above)
Comment (hardik01shah) [reply notebooks/jpldd_linedet_testbed.py]: done

File: notebooks/jpldd_linedet_testbed.py (hunk 5 above)
Comment (hardik01shah) [reply notebooks/jpldd_linedet_testbed.py]: done

**Comment by hardik01shah** [comment]
@RaphaelKreft made the changes you requested. Awaiting approval for merge

**Review by RaphaelKreft — APPROVED** [review]
lgtm

**Pull request merged**